src/dataframe/dataframe.jl

```diff
@@ -1223,34 +1236,59 @@ Base.convert(::Type{DataFrame}, A::AbstractMatrix) = DataFrame(A)
 
 Base.convert(::Type{DataFrame}, d::AbstractDict) = DataFrame(d, copycols=false)
 
-function Base.push!(df::DataFrame, row::Union{AbstractDict, NamedTuple}; columns::Symbol=:equal)
-    if !(columns in (:equal, :intersect))
-        throw(ArgumentError("`columns` keyword argument must be `:equal` or `:intersect`"))
+function Base.push!(df::DataFrame, row::Union{AbstractDict, NamedTuple}; cols::Symbol=:setequal,
+                    columns::Union{Nothing,Symbol}=nothing)
+    if !isnothing(columns)
+        cols = columns
+        Base.depwarn("`columns` keyword argument is deprecated. Use `cols` instead. ", :push!)
+    end
+    possible_cols = (:orderequal, :setequal, :intersect, :subset)
+    if !(cols in possible_cols)
+        throw(ArgumentError("`cols` keyword argument must be any of :" * join(possible_cols, ", :")))
     end
     nrows, ncols = size(df)
+    targetrows = nrows + 1
     if ncols == 0 && row isa NamedTuple
         for (n, v) in pairs(row)
             setproperty!(df, n, fill!(Tables.allocatecolumn(typeof(v), 1), v))
         end
         return df
     end
-    # Only check for equal lengths, as an error will be thrown below if some names don't match
-    if columns === :equal && length(row) != ncols
-        # TODO: add tests for this case after the deprecation period
-        Base.depwarn("In the future push! will require that `row` has the same number" *
-                      " of elements as is the number of columns in `df`. " *
-                      "Use `columns=:intersect` to disable this check.", :push!)
+    if cols === :orderequal
+        if row isa NamedTuple
+            if any(x -> x[1] != x[2], zip(propertynames(row), _names(df)))
+                throw(ArgumentError("all data frames to have the same column names " *
+                                    "and in the same order"))
+            end
+        end
+    elseif cols === :setequal || cols === :equal
+        if cols === :equal
+            Base.depwarn("`cols` value eqaual to `:equal` is deprecated." *
+                         "Use `:setequal` instead.", :push!)
+        end
+        # Only check for equal lengths if :setequal is selected,
+        # as an error will be thrown below if some names don't match
+        if length(row) != ncols
+            Base.depwarn("In the future `push!` with `cols` equal to `:setequal or " *
```

Comment:
There's a problem in that sentence AFAICT. :-)

Reply:
fixed